beluga_system_tests/test/test_system.cpp

```diff
@@ -41,405 +31,225 @@
-
- private:
-  // wrapped in a unique pointer to make it movable :)
-  std::unique_ptr<rosbag2_cpp::Reader> reader_;
-  std::size_t size_{0UL};
-};
 
-struct SE2dFromOdometryReader {
-  SE2dFromOdometryReader(const std::filesystem::path& bagfile_path, std::string_view topic_name)
-      : reader_{bagfile_path, std::move(topic_name)} {}
+  return ranges::views::generate_n([reader]() mutable { return reader->read_next<Message>(); }, size);
+}
 
-  Sophus::SE2d next() {
-    Sophus::SE2d ret;
-    tf2::convert(reader_.next().pose.pose, ret);
-    return ret;
-  }
+/// Transform a range of odometry messages into a range of SE2 elements.
+auto odometry_to_sophus() {
+  return ranges::views::transform([](const nav_msgs::msg::Odometry& message) {
+    Sophus::SE2d out;
+    tf2::convert(message.pose.pose, out);
+    return out;
+  });
+}
 
-  std::size_t size() { return reader_.size(); }
+// ****************************************************************************
+// Particle filter implementation variants.
+// ****************************************************************************
+
+// All possible particle filter configuration variants to test against.
+/**
+ * New particle filter implementation variants can be added by following these steps:
+ *
+ * - Create a new struct with the necessary configuration parameters (could be empty).
+ * - Add the new struct type to this variant declaration.
+ * - Add one or more instances of that struct to the `get_particle_filter_params` function.
+ * - Implement a new  `particle_filter_test` overload taking that struct as a first parameter.
+ */
+using ParticleFilterParams = std::variant<beluga_ros::AmclParams>;
+
+auto get_particle_filter_params() {
+  return std::vector<ParticleFilterParams>{
+      beluga_ros::AmclParams{},  // Default configuration.
+      []() {
+        auto params = beluga_ros::AmclParams{};
+        params.selective_resampling = true;  // Enable selective resampling.
+        return params;
+      }(),
+  };
+}
 
- private:
-  OneTopicReader<nav_msgs::msg::Odometry> reader_;
-};
+/// Overload of particle_filter_test with a standard AMCL implementation.
+template <class Map, class MotionModel, class SensorModel, class Distribution, class Range>
+void particle_filter_test(
+    const beluga_ros::AmclParams& params,
+    Map&& map,
+    MotionModel&& motion_model,
+    SensorModel&& sensor_model,
+    Distribution&& initial_distribution,
+    Range&& datapoints) {
+  auto filter = beluga_ros::Amcl{map, motion_model, sensor_model, params, std::execution::par};
+  filter.initialize(std::forward<Distribution>(initial_distribution));
+
+  // Tolerance values ​​prove that the filter performs approximately well, they do not prove accuracy.
+  // The values were adjusted to have a 100% success rate over 100 runs of the same test.
+  const double position_tolerance = 0.9;
+  const double orientation_tolerance = 30.0 * Sophus::Constants<double>::pi() / 180.0;
```

Comment:
@nahueespinosa nit: 
```suggestion
  constexpr double position_tolerance = 0.9;
  constexpr double orientation_tolerance = 30.0 * Sophus::Constants<double>::pi() / 180.0;
```

Reply:
`Sophus::Constants` are not constexpr :disappointed: https://github.com/strasdat/Sophus/blob/d270df2be6e46501b1e7efc09b107517e0be0645/sophus/common.hpp#L168